File: internal/web/statuses.go
package web

import (
	"encoding/json"
	"net/http"
	"strings"
)

// handleOutreachStatuses / handleApplicationStages read and write the two
// user-configurable status vocabularies (the dropdown labels for the jobs view's
// reply axis and application axis). GET -> {statuses: [...]}; PUT {statuses:
// [...]} replaces the list (empty/garbage rejected as 400). DB singletons, like
// the email template — no Runner involved.
func (s *Server) handleOutreachStatuses(w http.ResponseWriter, r *http.Request) {
	s.handleStatusList(w, r, s.DB.OutreachStatuses, s.DB.SetOutreachStatuses)
}

func (s *Server) handleApplicationStages(w http.ResponseWriter, r *http.Request) {
	s.handleStatusList(w, r, s.DB.ApplicationStages, s.DB.SetApplicationStages)
}

func (s *Server) handleStatusList(w http.ResponseWriter, r *http.Request, get func() ([]string, error), set func([]string) error) {
	switch r.Method {
	case http.MethodGet:
		list, err := get()
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"statuses": list})

	case http.MethodPut, http.MethodPost:
		var body struct {
			Statuses []string `json:"statuses"`
		}
		if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 8<<10)).Decode(&body); err != nil {
			http.Error(w, "invalid JSON: "+err.Error(), http.StatusBadRequest)
			return
		}
		if err := set(body.Statuses); err != nil {
			if strings.HasPrefix(err.Error(), "statuses ") {
				http.Error(w, err.Error(), http.StatusBadRequest)
				return
			}
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		list, err := get()
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"statuses": list})

	default:
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
	}
}

File: scout/web/routes/statuses.py
"""The two user-configurable status vocabularies (the jobs view dropdowns).

GET → {statuses}; PUT/POST {statuses}
replaces the list (empty/garbage rejected as 400 — the store's "statuses …"
ValueError → global 400). DB singletons, no Runner involved.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends
from starlette.responses import Response

from scout.store import statuses

from ..deps import get_db
from ..responses import json_error, json_response
from .core import decode_json, raw_body

router = APIRouter()


def _set_status_list(raw: bytes, con, get, set_) -> Response:
    body = decode_json(raw)
    lst = body.get("statuses")
    if lst is None:
        lst = []
    if not isinstance(lst, list) or not all(isinstance(x, str) for x in lst):
        return json_error("invalid JSON: statuses must be an array of strings", 400)
    set_(con, lst)  # validation → "statuses …" ValueError → global 400
    return json_response({"statuses": get(con)})


# --- outreach statuses (the reply axis) --------------------------------------


@router.get("/api/outreach-statuses")
def get_outreach_statuses(con=Depends(get_db)) -> Response:
    return json_response({"statuses": statuses.outreach_statuses(con)})


@router.api_route("/api/outreach-statuses", methods=["PUT", "POST"])
def set_outreach_statuses(raw: bytes = Depends(raw_body), con=Depends(get_db)) -> Response:
    return _set_status_list(raw, con, statuses.outreach_statuses, statuses.set_outreach_statuses)


# --- application stages (the application axis) -------------------------------


@router.get("/api/application-stages")
def get_application_stages(con=Depends(get_db)) -> Response:
    return json_response({"statuses": statuses.application_stages(con)})


@router.api_route("/api/application-stages", methods=["PUT", "POST"])
def set_application_stages(raw: bytes = Depends(raw_body), con=Depends(get_db)) -> Response:
    return _set_status_list(raw, con, statuses.application_stages, statuses.set_application_stages)
